### PR TITLE
Updates docs for Optax.

### PR DESCRIPTION
**Pull request by andsteing**
- Marks `flax.optim` as "effectively deprecated" in module pydoc (which is also shown on RTD).
- Marks `flax.training.lr_schedule` as "effectively deprecated" in module pydoc (which is also shown on RTD).
- Improves `flax.training.train_state` docs & includes them in RTD.
- Updates "Flax basics" Colab.
- Reinstates "Annotated MNIST" Colab, based on new MNIST source.

See #1357

### CODE REVIEW

File: flax/optim/__init__.py [hunk 1]
```diff
@@ -12,7 +12,17 @@
 # See the License for the specific language governing permissions and
 # limitations under the License.
 
-"""Flax Optimizer api."""
+"""Flax Optimizer api.
+
+Note that with `FLIP #1009`_ the optimizers in ``flax.optim`` were **effectively
+deprecated** in favor of Optax_ optimizers. There is not feature parity yet (e.g.
+``AdaFactor`` is missing in Optax), but the large majority of usecases is well
```
Comment (marcvanzee) [inline flax/optim/__init__.py:19]:
```suggestion
``AdaFactor`` is missing in Optax), but the large majority of use cases is well
```

File: flax/optim/__init__.py [hunk 2]
```diff
@@ -12,7 +12,17 @@
 # See the License for the specific language governing permissions and
 # limitations under the License.
 
-"""Flax Optimizer api."""
+"""Flax Optimizer api.
+
+Note that with `FLIP #1009`_ the optimizers in ``flax.optim`` were **effectively
+deprecated** in favor of Optax_ optimizers. There is not feature parity yet (e.g.
```
Comment (marcvanzee) [inline flax/optim/__init__.py:18]:
```suggestion
deprecated** in favor of Optax_ optimizers. There is no feature parity yet (e.g.
```

File: flax/optim/__init__.py [hunk 3]
```diff
@@ -12,7 +12,17 @@
...
+``AdaFactor`` is missing in Optax), but the large majority of usecases is well
+supported, and Optax is actively being developed and continuously adds new
+features and optimizers.
```
Comment (marcvanzee) [inline flax/optim/__init__.py:21]:
```suggestion
supported, and Optax is actively being developed and new features and
optimizers are being added continuously.
```

File: flax/training/train_state.py [hunk 4]
```diff
@@ -22,29 +22,30 @@
...
-      grads = grad_fn(state.params, batch)
-      state = state.apply_gradients(grads=grads)
+      state = TrainState.create(
+          apply_fn=model.apply,
+          params=variables['params'],
+          tx=tx)
+      grad_fn = jax.grad(make_loss_fn(state.apply_fn))
+      for batch in data:
+        grads = grad_fn(state.params, batch)
+        state = state.apply_gradients(grads=grads)
```
Comment (marcvanzee) [inline flax/training/train_state.py:34]: I don't think you need 4 indents, 2 indents and the double colon with an empty newline should be sufficient to get code view.

**Review by marcvanzee — CHANGES_REQUESTED** [review]
Why do we remove `docs/annotated_mnist.md` and we add `docs/notebooks/annotated_mnist.ipynb`?

File: flax/training/train_state.py (hunk 4 above)
Comment (andsteing) [reply flax/training/train_state.py:34]: Correct. Updated

**Comment by andsteing** [comment]
> Why do we remove `docs/annotated_mnist.md` and we add `docs/notebooks/annotated_mnist.ipynb`?

Because this allows users to open the code in a Colab, inspect & modify it. It's also a lot easier to keep it working by "running all" after changes. In a markdown file it would be laborious copy'n'paste to make sure the code is correct.

Note that I tried to use `jupytext` (as [does JAX](https://jax.readthedocs.io/en/latest/developer.html#syncing-notebooks)) but I couldn't get it to work, so I went for a good old Colab.

The content is exactly the same.

What would be the advantage of a markdown file over a Colab?

**Comment by andsteing** [comment]
Should also fix #659

**Comment by marcvanzee** [comment]
> What would be the advantage of a markdown file over a Colab?

It is a bit easier to read from source and update directly. I just was confused since the change doesn't really seem to be related with the description in the PR, but it makes sense.

**Review by marcvanzee — APPROVED** [review]

**Comment by andsteing** [comment]
> It is a bit easier to read from source and update directly. I just was confused since the change doesn't really seem to be related with the description in the PR, but it is fine.

Maybe I'll try again to update our docs to use `jupytext` then we would have the best from both worlds (i.e. simple text edits in markdown, but using Jupyter for code edits to make sure everything works as expected).

**Comment by 8bitmp3** [comment]
+1 jupytext @andsteing
